internal/service/wallet.go

```diff
@@ -199,30 +212,138 @@ func (s *walletService) GetWalletTransactions(ctx context.Context, walletID stri
+	var referenceType types.WalletTxReferenceType
+	var referenceID string
+
+	var idempotencyKey *string
+	if lo.FromPtr(req.IdempotencyKey) != "" {
+		idempotencyKey = req.IdempotencyKey
+	} else {
+		idempotencyKey = lo.ToPtr(types.GenerateUUID())
+	}
+
+	if err := req.Validate(); err != nil {
+		return nil, ierr.WithError(err).
+			WithHint("Invalid top up wallet request").
+			Mark(ierr.ErrValidation)
+	}
+
+	// this case is for free credits
+	if req.TransactionReason == types.TransactionReasonFreeCredit {
+		referenceType = types.WalletTxReferenceTypeExternal
+		referenceID = lo.FromPtr(idempotencyKey)
+	}
+
+	// this case is for when credits are purchased directly without invoice
+	if req.TransactionReason == types.TransactionReasonPurchasedCreditDirect {
+		referenceType = types.WalletTxReferenceTypeExternal
+		referenceID = lo.FromPtr(idempotencyKey)
+	}
+
+	// this case is for when credits are purchased and flexprice system generates an invoice and payment
+	if req.TransactionReason == types.TransactionReasonPurchasedCreditInvoiced {
+		referenceType = types.WalletTxReferenceTypeExternal
+		referenceID = lo.FromPtr(idempotencyKey)
+
+		// create services
+		invoiceService := NewInvoiceService(s.ServiceParams)
+		customerService := NewCustomerService(s.ServiceParams)
+		paymentService := NewPaymentService(s.ServiceParams)
+
+		// get customer data from wallet
+		w, err := s.WalletRepo.GetWalletByID(ctx, walletID)
+		if err != nil {
+			return nil, err
+		}
+
+		// get customer data from wallet
+		customer, err := customerService.GetCustomer(ctx, w.CustomerID)
+		if err != nil {
+			return nil, err
+		}
+
+		dueDate := time.Now().UTC()
+		displayName := "Purchased Credits"
+
+		err = s.DB.WithTx(ctx, func(ctx context.Context) error {
+
+			// create invoice with payment status pending
+			invoiceData := &dto.CreateInvoiceRequest{
+				CustomerID:     customer.ID,
+				AmountDue:      req.Amount,
+				Currency:       w.Currency,
+				InvoiceType:    types.InvoiceTypeCredit,
+				DueDate:        &dueDate,
+				IdempotencyKey: idempotencyKey,
+				InvoiceStatus:  lo.ToPtr(types.InvoiceStatusFinalized),
+				LineItems: []dto.CreateInvoiceLineItemRequest{
+					{
+						Amount:      req.Amount,
+						Quantity:    decimal.NewFromInt(1),
+						DisplayName: &displayName,
+					},
+				},
+				PaymentStatus: lo.ToPtr(types.PaymentStatusPending),
+			}
+
+			invoice, err := invoiceService.CreateInvoice(ctx, *invoiceData)
+			if err != nil {
+				return err
+			}
+			s.Logger.Infof("created invoice: %+v", invoice)
+
+			paymentData := &dto.CreatePaymentRequest{
+				IdempotencyKey:    lo.FromPtr(idempotencyKey),
+				DestinationType:   types.PaymentDestinationTypeInvoice,
+				DestinationID:     invoice.ID,
+				PaymentMethodType: types.PaymentMethodTypeOffline,
+				Amount:            req.Amount,
+				Currency:          w.Currency,
+				ProcessPayment:    false,
+			}
```

Comment:
set this true and make sure this marks payment success and do not manually update payment status